components/playlist/resources/components/app.tsx

```diff
@@ -0,0 +1,179 @@
+
+interface Props {
+  actions: any
+  playlistData: Playlist.State
+}
+
+interface State {
+  playlist?: PlaylistMojo.Playlist
+  experimentalUrl: string
+}
+
+export class PlaylistPage extends React.Component<Props, State> {
+  constructor (props: Props) {
+    super(props)
+    this.state = { experimentalUrl: '' }
+    this.onClickDownloadVideo = this.onClickDownloadVideo.bind(this)
+  }
+
+  get actions () {
+    return this.props.actions
+  }
+
+  getCurrentPlaylist = () => {
+    // TODO(sko): get current playlist index from store
+    return this.props.playlistData.lists.at(0)
+  }
+
+  getImgSrc = (itemId: string) => {
+    return 'chrome://playlist-data/' + itemId + '/thumbnail/'
+  }
+
+  getMediaSrc = (itemId: string) => {
+    return 'chrome://playlist-data/' + itemId + '/media'
+  }
+
+  get lazyButtonStyle () {
+    const lazyButtonStyle: any = {
+      alignItems: 'center',
+      WebkitAppearance: 'none',
+      width: '50px',
+      height: '50px',
+      display: 'flex',
+      borderRadius: '4px'
+    }
+    return lazyButtonStyle
+  }
+
+  getPlaylistHeader = (): Cell[] => {
+    return [
+      { content: 'INDEX' },
+      { content: 'NAME' },
+      { content: 'STATUS' },
+      { content: 'REMOVE' }
+    ]
+  }
+
+  getPlaylistRows = (playlist?: any): Row[] | undefined => {
+    if (!playlist || !playlist.items) {
+      return
+    }
+
+    return playlist.items.map((item: PlaylistMojo.PlaylistItem, index: any): any => {
+      const cell: Row = {
+        content: [
+          { content: (<div style={{ textAlign: 'center' }}>{index + 1}</div>) },
+          {
+            content: (
+                <PlaylistItem id={item.id} name={item.name} onClick={this.onClickItem.bind(this)}
+                    thumbnailUrl={item.thumbnailPath.url.startsWith('http') ? '' : this.getImgSrc(item.id)}/>
+            )
+          },
+          { content: (<span>{item.ready ? 'Ready' : 'Downloading'}</span>) },
+          { content: (<button style={this.lazyButtonStyle} onClick={this.onClickRemoveItemButton.bind(this, item.id)}><CloseCircleOIcon /></button>) }
+        ]
+      }
+      return cell
+    })
+  }
+
+  onClickRemoveItemButton = (playlistItemId: string) => {
+    getPlaylistAPI().removeItemFromPlaylist('', playlistItemId)
+  }
+
+  get pageHasDownloadableVideo () {
+    return this.state.experimentalUrl.startsWith('https://www.youtube.com/watch')
+  }
+
+  onChangeExperimentalUrl = (event: React.ChangeEvent<HTMLTextAreaElement>) => {
+    this.setState({ experimentalUrl: event.target.value })
+  }
+
+  onClickDownloadVideo = () => {
+    getPlaylistAPI().addMediaFilesFromPageToPlaylist('', this.state.experimentalUrl)
+  }
+
+  onClickItem = (itemId: string) => {
+    const item = this.getCurrentPlaylist()?.items.find((item: PlaylistMojo.PlaylistItem) => {
+      return item.id === itemId
+    })
+
+    if (!item || item.mediaPath.url.startsWith('http')) {
```

Comment:
Actually, we can load video before local cache is finished. But not sure how long the url is valid. The url looks time-limited but Not sure.

Reply:
I think checking `ready` property is more better?